doc/design/logging.md

```diff
@@ -0,0 +1,361 @@
+   call.
+
+4. Additionally, the minimal log level should be configurable by a compile time
+   switch and everything below this log level should be compiled to no-ops. With
+   this, developer can make excessive use of the most verbose log level and
+   disable it by default. When debugging, this can be turned on and help to find
+   the root of the problem.
+
+5. There shall also be the possibility to alter the behavior of the logger via
+   environment variables. This can be functionality like setting the log level or
+   deciding whether the timestamps or file names shall be part of a log message.
+
+6. Since iceoryx is often used as transport layer in a higher level framework, it
+   shall be possible to replace the default logger at runtime and delegate the
+   log messages to the logger of such a framework. To prevent misuse, this can
+   only be done up to a specific time. Afterwards an error message shall be
+   logged in the new and old logger and the error handler shall be called with a
+   moderate error level.
+
+7. The default logger can be replaced via the platform abstraction in order to
+   use the dedicated logger for log messages which are emitted before it could
+   be replaced at runtime, e.g. from global objects before `main`.
+
+8. Additionally, there shall be a compile time option to forward all log messages
+   independent of the log level threshold for cases where the logger framework
+   does the filtering itself.
+
+9. For tests the default logger shall be replaced to suppress the log messages in
+   passed tests but output them for failed tests. This keeps the noise low but
+   still provides useful information when needed. Additionally there shall be an
+   environment variable to circumvent the suppression when desired.
+
+### Solution
+
+#### Class diagram
+
+![logging class diagram](../website/images/logging_classes.svg)
+
+#### Logging with LogStream
+
+![logging with logstream](../website/images/logging_with_logstream.svg)
+
+#### Macro with lazy evaluation
+
+The buildup of the log message is only done when the log level is above the output
+threshold. This is accomplished by a macro with an incomplete if-statement.
+
+```cpp
+#define LAZY() if (cond)
+
+LAZY() expensiveFunctionCall();
+```
+
+In the example above `expensiveFunctionCall` is only executed when `cond` is `true`.
+If `cond` is a compile time constant set to `false`, the whole statement is compiled
+to a no-op and optimized away.
+
+This is the log macro with lazy evaluation
+
+```cpp
+#define IOX_LOG_INTERNAL(file, line, function, level)          \
+    if ((level) <= iox::log::Logger::minimalLogLevel()         \
+        && (iox::log::Logger::ignoreLogLevel()                 \
+            || (level) <= iox::log::Logger::getLogLevel()))    \
+    iox::log::LogStream(file, line, function, level).self()
+
+#define IOX_LOG(level) IOX_LOG_INTERNAL(__FILE__, __LINE__, __FUNCTION__, iox::log::LogLevel::level)
```

Comment:
I'll leave this to decide for later. The main question is if this struct should be part of the logger of live somewhere else since it might also be helpful for `Expects`/`Ensures`

Reply:
But if it would be helpful for `Expects`/`Ensures` you would be unable to use it since the requirement is that the logger has no dependency to `hoofs`.

I think this struct would be helpful in `hoofs` but could only be defined in the logger since this is our lowest building block.

So what do you think, does it make sense to introduce such a struct here and use it?

Reply:
Well, that was my point. It is not user facing and we could make use of it at several places but the question is where to put it so that we can use it everywhere. Therefore I would not do it in the initial implementation but wait for the error handler refactoring and then decide what to do.

Reply:
Expects/Ensures can be made obsolete but yes, we can bundle this later. I do not know yet where this should live. It is not critical, just saves us some boilerplate here and there while implementing the logging.

Reply:
@MatthiasKillat I'll use the implementation from the refactored error handler once it is available